packages/server/src/unstable-core-do-not-import/parser.ts

```diff
@@ -18,6 +20,13 @@ export type ParserArkTypeEsque<TInput, TParsedInput> = {
   infer: TParsedInput;
 };
 
+export type ParserStandardSchemaEsque<TInput, TParsedInput> = {
+  '~types'?: {
+    input: TInput;
+    output: TParsedInput;
+  };
```

Comment:
We have some different wip-branches where we do infer errors too, does `~types` also have an `Error`-subclass?

In the case of zod, we could infer that pretty easily in #6027 where we played with this, see L12 here:

https://github.com/trpc/trpc/blob/de219a3c4aab74cd2682f408c88f884be9cef587/packages/server/src/%40trpc/parser.ts#L2-L15

Reply:
What's the benefit of inferring the error type?

Reply:
Think e.g. a form submission - it'd be nice to infer the right path when rendering the errors

Reply:
Why do you need the error class for this?

Reply:
Cause there's no way to get the type of an error cause, so if you'd to `throw new Error("", { cause: issues })` you'd need a cast like `(err.cause as SchemaIssues<TSchema>).fieldErrors.email` for example. Or how would you infer here?

Would be nice if something strictly typed came out of the validate method directly as a built-in

Reply:
Are Zod's errors strictly typed? I thought it is the same error type for every schema? Zod's issues are also not typed as far as I know. But maybe I'm wrong. Could the paths also be inferred from the input type?

Reply:
they are: 
![CleanShot 2024-10-08 at 00 24 15](https://github.com/user-attachments/assets/ee4adf11-018a-4d76-bfac-231644dfb651)

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJvgNwBQFAxhAHYDO8IAngMrUAWApiAIZwAvCgB0EAEYArbtRgAKJJggQAXKKZRgdAOZyAlABo44vlDXIRdAK4hx3KPvR6qtRvCjcGVgDbxhrDh5+EQY+TG4ABVMGbgU4OggYADFlNQByNKMTMzgANgBOJyoPL18RexwoCgB6arh6+oA9AH4KEp8YcqhKkUxvPhgYbjp9XuBub2IAUW7oBhEgA

Reply:
Thank you for the example! I understand it much better now. I think this is also possible with Standard Schema. We know the input and output type, so we know all possible paths that can occur, and can transform them into an object of dot paths. I can provide example code.

Reply:
The idea is basically to infer the input type from the schema. Standard Schema provides an `InferInput` util type for this. Then we could use a modified version of [ts-essentials'](https://github.com/ts-essentials/ts-essentials) `Paths` type, which converts the input type into a union of all possible paths. Finally, we pass this type to `Record` to construct the `FieldErrors` type.

```ts
type FieldErrors = Record<Paths<InferInput<TSchema>>, string[]>
```

Reply:
Using an object might not be the best idea as errors could occur on e.g both `foo` and `foo.bar` simultaneously - an array with issues and paths is probably best for a generic lib - you can supply functions to turn it into an object

Reply:
> Using an object might not be the best idea as errors could occur on e.g both foo and foo.bar simultaneously

You don't think this is a problem, or are you aware of an edge case? That's why I wrote that the `Paths` type needs to be modified. I can provide a raw implementation if you are interested.

> an array with issues and paths is probably best for a generic lib

I completely agree. That's exactly what the Standard Schema does.

> you can supply functions to turn it into an object

I have thought about providing such a function with `@standard-schema/utils`, but such an implementation requires a lot of thought to make it work for everyone. Some libraries are probably only interested in nested messages, because they assume (or know) that the root type is always valid. Others are also interested in errors related to the root schema. There may also be edge cases where a path cannot be converted to a dot path. Then there are libraries that need the value to be a `string` and others that need it to be a `string[]`.

We are very interested in and grateful for feedback from libraries like tRPC to make the spec as perfect as possible.